python/pylibcudf/tests/test_lists.py

```diff
@@ -161,19 +161,37 @@ def test_reverse(list_column):
     assert_column_eq(expect, got)
 
 
-# https://github.com/rapidsai/cudf/issues/19346
-@pytest.mark.flaky(reruns=5)
-def test_segmented_gather(test_data):
-    list_column1, list_column2 = test_data[0]
-
-    plc_column1 = plc.Column.from_arrow(pa.array(list_column1))
-    plc_column2 = plc.Column.from_arrow(pa.array(list_column2))
+def test_segmented_gather():
+    source_column = plc.Column.from_arrow(
+        pa.array([[10, 20], [], [30], [None, 50, 30, 20]])
+    )
+    gather_map_list = plc.Column.from_arrow(
+        pa.array([[1, 0, 1], [], [], [0, 1, 3]])
+    )
+    got = plc.lists.segmented_gather(source_column, gather_map_list)
+    expect = pa.array([[20, 10, 20], [], [], [None, 50, 20]])
 
-    got = plc.lists.segmented_gather(plc_column2, plc_column1)
+    assert_column_eq(expect, got)
 
-    expect = pa.array([[8, 9], [14], [0], [0, 0]])
 
-    assert_column_eq(expect, got)
+def test_segmented_gather_out_of_bounds():
+    source_column = plc.Column.from_arrow(pa.array([[], [10, 20]]))
+    gather_map_list = plc.Column.from_arrow(
+        pa.array(
+            [[0], [0, 2]],
+        )
+    )
+    got = plc.lists.segmented_gather(source_column, gather_map_list)
+    result = plc.interop.to_arrow(got)
+    # indexing past the interior lists is undefined behavior,
+    # so the only things we can check are the general shape of the output
+
+    assert got.null_count() == 0
+    assert got.size() == 2
+    assert len(result[0]) == 1
+    assert len(result[1]) == 2
+    # this is the only indexer that isn't out of bounds.
+    assert result[1][0].as_py() == 10
```

Comment:
We should add a `bounds_policy` argument to `segmented_gather` in pylibcudf. And then test in and out of bounds in the same test in `test_segmented_gather`.

See this example https://github.com/rapidsai/cudf/blob/17e53b18969194740522f01761a12ff19bfdc105/python/pylibcudf/pylibcudf/copying.pyx#L56-L93
```suggestion
```

Reply:
I'm fine with deleting this test now and adding the `bounds_policy` in a follow-up.

Reply:
I think you can check the actual values because they are nullified. Eg.
```
source_column       : [{"a", "b", "c", "d"}, {"1", "2", "3", "4"}, {"x", "y", "z"}]
gather_map_list     : [{0, -1, 4, -5}, {1, 3, 5}, {}]
 
result_with_nullify : [{"a", "d", null, null}, {"2", "4", null}, {}]

```

Reply:
Sorry I was talking about the test with `DONT_CHECK`. For those, we can't say what the out-of-bounds values will be (even whether they're null or not).

Agreed that with `NULLIFY` we can and should check the nulls. My [new test](https://github.com/rapidsai/cudf/pull/19411/files#diff-0d10375f22ead8483d6a449e85eb72924f5d38e2e8bde1c18701a2c4614dec06R197) here does that.